src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs

```diff
@@ -126,16 +127,34 @@ public CompilationResult Compile([NotNull] IFileInfo fileInfo, [NotNull] string
             }
         }
 
-        private List<MetadataReference> GetApplicationReferences()
+        private IReadOnlyList<MetadataReference> GetApplicationReferences()
         {
             var references = new List<MetadataReference>();
 
-            var export = _libraryManager.GetAllExports(_environment.ApplicationName);
+            // Get the MetadataReference for the executing application. If it's a Roslyn reference,
+            // we can copy the references created when compiling the application to the Razor page being compiled.
+            // This avoids performing expensive calls to MetadataReference.CreateFromImage.
+            var libraryExport = _libraryManager.GetLibraryExport(_environment.ApplicationName);
+            if (libraryExport.MetadataReferences.Count > 0)
+            {
+                Debug.Assert(libraryExport.MetadataReferences.Count == 1,
+                             "Expected 1 MetadataReferences, found " + libraryExport.MetadataReferences.Count);
+                var roslynReference = libraryExport.MetadataReferences[0] as IRoslynMetadataReference;
+                if (roslynReference != null)
+                {
+                    var compilationReference = (CompilationReference)roslynReference.MetadataReference;
+
+                    references.AddRange(compilationReference.Compilation.References);
+                    references.Add(roslynReference.MetadataReference);
 
+                    return references;
+                }
+            }
+
+            var export = _libraryManager.GetAllExports(_environment.ApplicationName);
             foreach (var metadataReference in export.MetadataReferences)
             {
-                // Taken from https://github.com/aspnet/KRuntime/blob/757ba9bfdf80bd6277e715d6375969a7f44370ee/src/...
-                // Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs#L164
+                // Taken from https://github.com/aspnet/KRuntime/blob/757ba9bfdf80bd6277e715d6375969a7f44370ee/src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs#L164
```

Comment:
ick.  can you reference the file directly (shortening the URI) then mention the specific commit separately?